atom/renderer/atom_render_frame_observer.cc

```diff
@@ -99,4 +142,78 @@ bool AtomRenderFrameObserver::ShouldNotifyClient(int world_id) {
     return IsMainWorld(world_id);
 }
 
+bool AtomRenderFrameObserver::OnMessageReceived(const IPC::Message& message) {
+  bool handled = true;
+  IPC_BEGIN_MESSAGE_MAP(AtomRenderFrameObserver, message)
+    IPC_MESSAGE_HANDLER(AtomViewMsg_Message, OnBrowserMessage)
+    IPC_MESSAGE_HANDLER(AtomViewMsg_Offscreen, OnOffscreen)
+    IPC_MESSAGE_UNHANDLED(handled = false)
+  IPC_END_MESSAGE_MAP()
+
+  return handled;
+}
+
+void AtomRenderFrameObserver::OnOffscreen() {
+  blink::WebView::SetUseExternalPopupMenus(false);
+}
+
+void AtomRenderFrameObserver::OnBrowserMessage(bool send_to_all,
+                                              const base::string16& channel,
+                                              const base::ListValue& args) {
+  // Don't handle browser messages before document element is created.
+  // When we receive a message from the browser, we try to transfer it
+  // to a web page, and when we do that Blink creates an empty
+  // document element if it hasn't been created yet, and it makes our init
+  // script to run while `window.location` is still "about:blank".
+  if (!document_created_)
+    return;
+
+  if (!render_frame()->GetRenderView()->GetWebView())
+    return;
+
+  blink::WebFrame* frame = render_frame()->GetRenderView()->GetWebView()->MainFrame();
```

Comment:
atom/renderer/atom_render_frame_observer.cc:174:  Lines should be <= 80 characters long  [whitespace/line_length] [2]